**Задание №1**

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

data = pd.DataFrame()
for i in range(10):
  info = pd.read_csv(f"/content/Students_info_{i}.csv")
  marks = pd.read_csv(f"/content/Students_marks_{i}.csv")
  merged = pd.merge(info, marks, on = 'index')
  data = pd.concat([data,merged])
data.head()

index  gender race/ethnicity parental level of education         lunch  \
0      0  female        group B           bachelor's degree      standard   
1      1  female        group C                some college      standard   
2      2  female        group B             master's degree      standard   
3      3    male        group A          associate's degree  free/reduced   
4      4    male        group C                some college      standard   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75

**Задание №2**

In [2]:
data = data.drop('index', axis = 1)
data.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75

**Задание №3**

In [3]:
data.shape

(1000, 9)

**Задание №4**

In [4]:
data.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

**Задание №5**

In [5]:
data.isnull().any().any()

False

**Задание №6**

In [6]:
data[data.columns[6:9]].mean()

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

**Задание №7**

In [7]:
mean_scores = data.groupby('test preparation course' ).mean(numeric_only=True)[['math score', 'reading score', 'writing score']]
mean_scores.head()


math score  reading score  writing score
test preparation course                                          
completed                 69.695531      73.893855      74.418994
none                      64.077882      66.534268      64.504673

Из полученных данных можно заметить, что баллы за эказмены студентов, которые проходили курс, выше, чем у студентов, которые этот подготовительный курс не прошли.

**Задание №8**

In [8]:
data['lunch'].unique()

array(['standard', 'free/reduced'], dtype=object)

**Задание №9**

In [9]:
data = data.rename(columns = {'parental level of education':'education', 'test preparation course': 'test preparation'})
data.head()

gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   

    group  math score  reading score  writing score  
0  group1          72             72             74  
1  group1          69             90             88  
2  group1          90             95             93  
3  group1          47             57             44  
4  group1          76             78             75

**Задание №10**

In [10]:
passmark = 50
data['math_pass'] = data['math score'] >= passmark
print(data['math_pass'].mean())
ans2 = data.groupby('test preparation')['math_pass'].mean()
print(ans2['completed'])
ans3 = data[(data['gender'] == 'female') & (data['test preparation'] == 'none')]
print(1 - ans3['math_pass'].mean())

0.865
0.9217877094972067
0.20958083832335328


Итого имеем:

*   0.865 студентов сдали экзамен по математике
*   0.921 сдуентов прошли курс подготовки к эказмену и сдали экзамен по математике
*   0.2 женщин не прошли курс подготовки и не сдали экзамен по математике






**Задание №11**

In [11]:
import time
start_time1 = time.time()
ans1 = data.groupby('race/ethnicity')['reading score'].mean()
print(ans1)
time_ans1 = time.time() - start_time1
print(time_ans1)

race/ethnicity
group A    64.674157
group B    67.352632
group C    69.103448
group D    70.030534
group E    73.028571
Name: reading score, dtype: float64
0.006410360336303711


In [12]:
import time
start_time2 = time.time()
ans2 = data.groupby('education')['writing score'].min()
print(ans2)
time_ans2 = time.time() - start_time2
print(time_ans2)

education
associate's degree    35
bachelor's degree     38
high school           15
master's degree       46
some college          19
some high school      10
Name: writing score, dtype: int64
0.008520841598510742


**Задание №12**

In [13]:
import time
start_time3 = time.time()
ethnics = data['race/ethnicity'].unique()
reading_avg = {}
for group in ethnics:
  avg = data[data['race/ethnicity'] == group]['reading score'].mean()
  reading_avg[group] = avg
print(reading_avg)
print(time.time() - start_time3)
print(time.time() - start_time3 - time_ans1)


{'group B': 67.35263157894737, 'group C': 69.10344827586206, 'group A': 64.67415730337079, 'group D': 70.03053435114504, 'group E': 73.02857142857142}
0.008601665496826172
0.0029058456420898438


Разница во времени в первой задаче составила ***0.002340 секунды***

In [14]:
import time
start_time4 = time.time()
educations = data['education'].unique()
educ_min = {}
for level in educations:
  min = data[data['education'] == level]['writing score'].min()
  educ_min[level] = min
print(educ_min)
print(time.time() - start_time4)
print(time.time() - start_time4 - time_ans2)

{"bachelor's degree": 38, 'some college': 19, "master's degree": 46, "associate's degree": 35, 'high school': 15, 'some high school': 10}
0.008407115936279297
0.0026803016662597656


Разница во времени для второй задачи составила ***0.00566411 секунды***

**Задание №13**

In [17]:
pv_table = data.pivot_table(index = ['gender', 'education'],
                            values = ['math score', 'reading score','writing score'],
                            aggfunc = 'mean')
print(pv_table)

                           math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000


**Задание №14**

In [18]:
passmark = 50
data['Math_PassStatus'] = data['math score'].apply(lambda x: 'P' if x >= passmark else 'F')
data['Reading_PassStatus'] = data['reading score'].apply(lambda x: 'P' if x >= passmark else 'F')
data['Writing_PassStatus'] = data['writing score'].apply(lambda x: 'P' if x >= passmark else 'F')
print(data['Writing_PassStatus'].value_counts())
print(data['Reading_PassStatus'].value_counts())
print(data['Writing_PassStatus'].value_counts())

P    886
F    114
Name: Writing_PassStatus, dtype: int64
P    910
F     90
Name: Reading_PassStatus, dtype: int64
P    886
F    114
Name: Writing_PassStatus, dtype: int64


**Задание №15**

In [19]:
data['OverAll_PassStatus'] = data.apply(lambda x: 'F' if x['Math_PassStatus'] == 'F' or x['Reading_PassStatus'] == 'F' or x['Writing_PassStatus'] == 'F' else 'P', axis = 1 )
print(data['OverAll_PassStatus'].value_counts())

P    812
F    188
Name: OverAll_PassStatus, dtype: int64


**Задание №16**

In [20]:
def GetGrade(average_mark):
  if average_mark > 90:
    return 'A'
  elif average_mark >= 80:
    return 'B'
  elif average_mark >= 70:
    return 'C'
  elif average_mark >= 60:
    return 'D'
  elif average_mark >= 50:
    return 'E'
  else:
    return 'F'
data['average_mark'] = (data['math score'] + data['reading score'] + data['writing score']) / 3
data['Grade'] = data['average_mark'].apply(GetGrade)
data = data.drop('average_mark', axis = 1)
data['Grade'].value_counts()

C    261
D    256
E    182
B    148
F    103
A     50
Name: Grade, dtype: int64